File: python/binarySearchTree/orderStatistics.py
from node import Node

class OrderStatistics:
    def bfsPrint(self, node, key='key'):
        result = []
        q = []
        q.append(node)

        while(len(q) > 0):
            tmp = q.pop(0)
            result.append(str(getattr(tmp, key)))

            if tmp.left is not None:
                q.append(tmp.left)
            if tmp.right is not None:
                q.append(tmp.right)
        
        return ','.join(result)

    def find(self, key, node):
        if key == node.key:
            return node

        if key > node.key:
            if node.right is None:
                return node
            return self.find(key, node.right)
        else:
            if node.left is None:
                return node
            return self.find(key, node.left)

    def insert(self, key, root):
        if root is None:
            return Node(key, None)

        node = self.find(key, root)
        new = Node(key, node)

        if key > node.key:
            node.right = new
        else:
            node.left = new

        return root

    def avlInsert(self, key, root):
        root = self.insert(key, root)
        node = self.find(key, root)

        return self.rebalance(node)

    def rebalance(self, node):
        if self.getHeight(node.left) > self.getHeight(node.right) + 1:
            self.rebalanceRight(node)
        
        if self.getHeight(node.right) > self.getHeight(node.left) + 1:
            self.rebalanceLeft(node)

        self.adjustHeightSize(node)

        p = node.parent
        if p is not None:
            return self.rebalance(p)
        
        return node

    def rebalanceRight(self, node):
        l = node.left
        if self.getHeight(l.right) > self.getHeight(l.left):
            self.rotateLeft(l)
        self.rotateRight(node)

    def rebalanceLeft(self, node):
        r = node.right
        if self.getHeight(r.left) > self.getHeight(r.right):
            self.rotateRight(r)
        self.rotateLeft(node)

    def rotateRight(self, node):
        p = node.parent
        l = node.left
        r = l.right

        if p is not None:
            if p.left is not None and p.left.key == node.key:
                p.left = l
            if p.right is not None and p.right.key == node.key:
                p.right = l
        l.parent = p

        l.right = node
        node.parent = l

        node.left = r
        if r is not None:
            r.parent = node

        self.adjustHeightSize(node) # old root
        self.adjustHeightSize(l) # new root

    def rotateLeft(self, node):
        p = node.parent
        r = node.right
        l = r.left

        if p is not None:
            if p.left is not None and p.left.key == node.key:
                p.left = r
            if p.right is not None and p.right.key == node.key:
                p.right = r
        r.parent = p

        r.left = node
        node.parent = r

        node.right = l
        if l is not None:
            l.parent = node

        self.adjustHeightSize(node) # old root
        self.adjustHeightSize(r) # new root

    def getHeight(self, node):
        return 0 if node is None else node.height
    
    def getSize(self, node):
        return 0 if node is None else node.size
    
    def adjustHeightSize(self, node):
        node.height = max(self.getHeight(node.left), self.getHeight(node.right)) + 1
        node.size = self.getSize(node.left) + self.getSize(node.right) + 1

    def orderStatistic(self, root, n):
        size = self.getSize(root.left)

        if n == size + 1:
            return root
        elif n < size + 1:
            return self.orderStatistic(root.left, n)
        elif n > size + 1:
            return self.orderStatistic(root.right, n - size - 1)

    def reverseOrderStatistic(self, root, key):
        if key < root.key:
            return self.reverseOrderStatistic(root.left, key)
        else:
            tmp = 1 + self.getSize(root.left)
            if key == root.key:
                return tmp
            else:
                return tmp + self.reverseOrderStatistic(root.right, key)

File: java/src/binarySearchTree/OrderStatistics.java
package binarySearchTree;

import java.util.Queue;
import java.util.LinkedList;
import java.util.StringJoiner;

public class OrderStatistics {
    public Node find(int key, Node node) {
        if (key == node.key)
            return node;

        if (key < node.key) {
            if (node.left != null)
                return this.find(key, node.left);
            return node;
        } else {
            if (node.right != null)
                return this.find(key, node.right);
            return node;
        }
    }

    public Node insert(int key, Node root) {
        if (root == null)
            return new Node(key, null);

        Node node = this.find(key, root);
        Node newNode = new Node(key, node);

        if (key < node.key)
            node.left = newNode;
        else
            node.right = newNode;

        return root;
    }

    public Node avlInsert(int key, Node root) {
        root = this.insert(key, root);
        Node node = this.find(key, root);
        return this.rebalance(node);
    }

    public Node rebalance(Node node) {
        if (this.getHeight(node.left) > this.getHeight(node.right) + 1)
            this.rebalanceRight(node);

        if (this.getHeight(node.right) > this.getHeight(node.left) + 1)
            this.rebalanceLeft(node);

        this.adjustHeightAndSize(node);

        Node p = node.parent;
        if (p != null)
            return this.rebalance(p);
        else
            return node;
    }

    public int getHeight(Node node) {
        return node == null ? 0 : node.height;
    }

    public int getSize(Node node) {
        return node == null ? 0 : node.size;
    }

    public void rebalanceRight(Node node) {
        Node l = node.left;
        if (this.getHeight(l.right) > this.getHeight(l.left))
            this.rotateLeft(l);
        rotateRight(node);
    }

    public void rebalanceLeft(Node node) {
        Node r = node.right;
        if (this.getHeight(r.left) > this.getHeight(r.right))
            this.rotateRight(r);
        this.rotateLeft(node);
    }

    public void rotateRight(Node node) {
        Node p = node.parent;
        Node l = node.left;
        Node r = l.right;

        if (p != null) {
            if (p.left != null && p.left.key == node.key)
                p.left = l;
            if (p.right != null && p.right.key == node.key)
                p.right = l;
        }
        l.parent = p;

        l.right = node;
        node.parent = l;

        node.left = r;
        if (r != null)
            r.parent = node;

        this.adjustHeightAndSize(node);
        this.adjustHeightAndSize(l);
    }

    public void rotateLeft(Node node) {
        Node p = node.parent;
        Node r = node.right;
        Node l = r.left;

        if (p != null) {
            if (p.left != null && p.left.key == node.key)
                p.left = r;
            if (p.right != null && p.right.key == node.key)
                p.right = r;
        }
        r.parent = p;

        r.left = node;
        node.parent = r;

        node.right = l;
        if (l != null)
            l.parent = node;

        this.adjustHeightAndSize(node);
        this.adjustHeightAndSize(r);
    }

    public void adjustHeightAndSize(Node node) {
        node.height = Math.max(this.getHeight(node.left), this.getHeight(node.right)) + 1;
        node.size = this.getSize(node.left) + this.getSize(node.right) + 1;
    }

    public Node orderStatistic(Node root, int n) {
        int s = this.getSize(root.left);

        if (n == s + 1)
            return root;
        else if (n < s + 1)
            return this.orderStatistic(root.left, n);
        else
            return this.orderStatistic(root.right, n - s - 1);
    }

    public int reverseOrderStatistic(Node root, int key) {
        if (key < root.key) {
            return this.reverseOrderStatistic(root.left, key);
        } else {
            int n = this.getSize(root.left) + 1;

            if (key > root.key)
                return n + this.reverseOrderStatistic(root.right, key);
            else
                return n;
        }
    }

    public String bfsPrint(Node root, String type) {
        StringJoiner sj = new StringJoiner(",");
        Queue<Node> q = new LinkedList<Node>();
        q.add(root);

        while(!q.isEmpty()) {
            Node node = q.poll();
            sj.add(node.getDataByType(type));

            if (node.left != null)
                q.add(node.left);
            if (node.right != null)
                q.add(node.right);
        }

        return sj.toString();
    }
}
